lib/src/rules/avoid_unstable_final_fields.dart

```diff
@@ -0,0 +1,685 @@
+// Copyright (c) 2022, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/token.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/element.dart';
+import 'package:analyzer/diagnostic/diagnostic.dart';
+import 'package:analyzer/src/diagnostic/diagnostic.dart';
+
+import '../analyzer.dart';
+
+const _desc = r'Avoid overriding a final field to return '
+    'different values if called multiple times';
+
+const _details = r'''
+
+**AVOID** overriding or implementing a final field as a getter which could
+return different values if it is invoked multiple times on the same receiver.
+This could occur because the getter is an implicitly induced getter of a
+non-final field, or it could be an explicitly declared getter with a body
+that isn't known to return the same value each time it is called.
+
+The underlying motivation for this rule is that if it is followed then a final
+field is an immutable property of an object. This is important for correctness
+because it is then safe to assume that the value does not change during the
+execution of an algorithm. In contrast, it may be necessary to re-check any
+other getter repeatedly if it is not known to have this property. Similarly,
+it is safe to cache the immutable property in a local variable and promote it,
+but for any other property it is necessary to check repeatedly that the
+underlying property hasn't changed since it was promoted.
+
+**BAD:**
+```dart
+class A {
+  final int i;
+  A(this.i);
+}
+
+var j = 0;
+
+class B1 extends A {
+  int get i => ++j + super.i; // LINT.
+  B1(super.i);
+}
+
+class B2 implements A {
+  int i; // LINT.
+  B2(this.i);
+}
+```
+
+**GOOD:**
+```dart
+class C {
+  final int i;
+  C(this.i);
+}
+
+class D1 implements C {
+  late final int i = someExpression; // OK.
+}
+
+class D2 extends C {
+  int get i => super.i + 1; // OK.
+  D2(super.i);
+}
+```
+
+''';
+
+class AvoidUnstableFinalFields extends LintRule {
+  AvoidUnstableFinalFields()
+      : super(
+            name: 'avoid_unstable_final_fields',
+            description: _desc,
+            details: _details,
+            group: Group.errors);
+
+  @override
+  void registerNodeProcessors(
+      NodeLintRegistry registry, LinterContext context) {
+    var visitor = _Visitor(this, context);
+    registry.addFieldDeclaration(this, visitor);
+    registry.addMethodDeclaration(this, visitor);
+  }
+}
+
+bool _isLocallyStable(Element element) {
+  if (element is PropertyAccessorElement &&
+      element.isSynthetic &&
+      element.correspondingSetter == null) {
```

Comment:
`correspondingSetter` will also return `null` if `element` is a setter. You probably want to first test `element.isGetter`.